### PR TITLE
Reduce locking contention per LibuvThread.Post

### PR DESCRIPTION
**Pull request by benaadams**
Use ConcurrentQueue for `Work` and `CloseHandle` items; many threads `Post`, should be low contention for single thread dequing.

Don't trigger `uv_async_send` syscall while the loop woken up and about to or processing that type of element; or on step prior.

Use empty struct `VoidResult` for `TaskCompletionSource<VoidResult>` as don't care about result; no need for a memory barrier write for `object`

### CODE REVIEW

**Comment by benaadams** [comment]
OSX
```
Failed   ConnectionResetMidRequestIsLoggedAsDebug
Error Message:
 Logged event should not have been higher than debug.
Expected: False
Actual:   True
Stack Trace:
   at Microsoft.AspNetCore.Server.Kestrel.FunctionalTests.RequestTests.<ConnectionResetMidRequestIsLoggedAsDebug>d__15.MoveNext() in /_/test/Kestrel.FunctionalTests/RequestTests.cs:line 462
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
Standard Output Messages:
 Debug: Connection id "0HLBHL33ILSS3" started.
 Debug: Connection id "0HLBHL33ILSS3" reset.
 Debug: Connection id "0HLBHL33ILSS3" disconnecting.
 Information: Connection id "0HLBHL33ILSS3" communication error.
 Debug: Connection id "0HLBHL33ILSS3" sending FIN.
 Debug: Connection id "0HLBHL33ILSS3" stopped.
```

File: src/Kestrel.Transport.Libuv/Internal/LibuvThread.cs [hunk 1]
```diff
@@ -309,80 +315,101 @@ private void ThreadStart(object parameter)
...
         }
 
-        private bool DoPostWork()
+        private void DoPostWork(int maxItems)
         {
-            Queue<Work> queue;
-            lock (_workSync)
+            var i = 0;
+            var queue = _work;
+            bool shouldLoop;
```
Comment (benaadams) [inline src/Kestrel.Transport.Libuv/Internal/LibuvThread.cs]: This should be
```csharp
bool shouldLoop = Volatile.Read(ref _workAlerted.Value) == 1;
while (shouldLoop)
```
but it makes the tests unhappy about clean termination...

Comment (benaadams) [reply src/Kestrel.Transport.Libuv/Internal/LibuvThread.cs]: Fixed

**Comment by davidfowl** [comment]
Any perf results to verify this does what it claims 😄 ?

**Comment by benaadams** [comment]
Two linux Standard D64 v3 battle...

Pre
```
  64 threads and 256 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     3.84ms   11.94ms 269.22ms   88.62%
    Req/Sec    44.70k    16.16k   99.02k    69.40%
  28792380 requests in 10.10s, 3.54GB read
Requests/sec: 2850821.02
Transfer/sec:    358.88MB

  64 threads and 1024 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     9.89ms   23.88ms 576.58ms   94.51%
    Req/Sec    55.16k    21.28k  236.51k    74.30%
  35003078 requests in 10.10s, 4.30GB read
Requests/sec: 3465630.70
Transfer/sec:    436.27MB

  64 threads and 4096 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    52.15ms  107.04ms   1.99s    90.73%
    Req/Sec    48.43k    24.34k  340.59k    70.66%
  31626097 requests in 10.10s, 3.89GB read
  Socket errors: connect 0, read 0, write 0, timeout 1
Requests/sec: 3131147.62
Transfer/sec:    394.16MB

  64 threads and 16384 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   246.84ms  343.47ms   2.00s    86.88%
    Req/Sec    41.20k    37.25k  365.30k    72.11%
  24531512 requests in 10.10s, 3.02GB read
  Socket errors: connect 0, read 138, write 0, timeout 4873
Requests/sec: 2428899.37
Transfer/sec:    305.76MB
```

Post
```
  64 threads and 256 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    14.90ms   23.81ms 329.58ms   88.74%
    Req/Sec     8.35k     2.03k   35.24k    74.29%
  5376579 requests in 10.10s, 676.83MB read
Requests/sec: 532355.06
Transfer/sec:     67.02MB

  64 threads and 1024 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    21.85ms   54.05ms 856.01ms   93.18%
    Req/Sec    34.40k    25.08k  199.77k    62.28%
  21578292 requests in 10.10s, 2.65GB read
Requests/sec: 2136509.40
Transfer/sec:    268.95MB

  64 threads and 4096 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    77.30ms  163.23ms   1.99s    92.62%
    Req/Sec    45.79k    24.18k  308.08k    66.73%
  28715331 requests in 10.10s, 3.53GB read
  Socket errors: connect 0, read 0, write 0, timeout 54
Requests/sec: 2844263.72
Transfer/sec:    358.05MB

  64 threads and 16384 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   219.31ms  307.53ms   2.00s    88.65%
    Req/Sec    26.65k    24.69k  322.97k    83.80%
  16916975 requests in 10.10s, 2.08GB read
  Socket errors: connect 0, read 34, write 0, timeout 6631
Requests/sec: 1674993.46
Transfer/sec:    210.86MB
```

😢

**Comment by benaadams** [comment]
Not value for change

**Pull request closed without merging**